### PR TITLE
Fixed #32285 -- Raised ImproperlyConfigured when AppConfig.label is not a valid Python identifier.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/32285

### CODE REVIEW

File: django/apps/config.py [hunk 1]
```diff
@@ -34,6 +34,10 @@ def __init__(self, app_name, app_module):
         # This value must be unique across a Django project.
         if not hasattr(self, 'label'):
             self.label = app_name.rpartition(".")[2]
+        elif not self.label.isidentifier():
+            raise ImproperlyConfigured(
+                "Invalid label '%s' for app '%s'" % (self.label, app_name)
```
Comment (felixxm) [inline django/apps/config.py:39]: `The app label '%s' is not a valid Python identifier.`

**Comment by hramezani** [comment]
@felixxm the following test fails
https://github.com/django/django/blob/5fcfe5361e5b8c9738b1ee4c1e9a6f293a7dda40/tests/migrations/test_state.py#L515-L533

Because we set the `label` here in `AppConfigStub` class:
https://github.com/django/django/blob/5fcfe5361e5b8c9738b1ee4c1e9a6f293a7dda40/django/db/migrations/state.py#L234

I don't know how to change the test :thinking:

**Comment by felixxm** [comment]
> I don't know how to change the test thinking

There is nothing to change in the test, it's a regression test for ticket-23483. Your patch must handle this case.

**Comment by hramezani** [comment]
> > I don't know how to change the test thinking
> 
> There is nothing to change in the test, it's a regression test for [ticket-23483](https://code.djangoproject.com/ticket/23483). Your patch must handle this case.

Thanks @felixxm for the explanation. I've changed my patch to handle this

File: django/db/migrations/state.py [hunk 2]
```diff
@@ -231,11 +231,11 @@ class AppConfigStub(AppConfig):
     path = ''
 
     def __init__(self, label):
-        self.label = label
         # App-label and app-name are not the same thing, so technically passing
         # in the label here is wrong. In practice, migrations don't care about
         # the app name, but we need something unique, and the label works fine.
-        super().__init__(label, None)
+        self.label = label
+        self.name = label
```
Comment (felixxm) [inline django/db/migrations/state.py:238]: I wonder does it still make sense to subclass `AppConfig`, we only want to have a dict of models for each app. Subclassing `AppConfig` looks unnecessary :thinking: @charettes Can I ask for your opinion?

File: django/apps/config.py [hunk 3]
```diff
@@ -34,6 +34,10 @@ def __init__(self, app_name, app_module):
         # This value must be unique across a Django project.
         if not hasattr(self, 'label'):
             self.label = app_name.rpartition(".")[2]
+        if not self.label.isidentifier():
+            raise ImproperlyConfigured(
+                "The app label '%s' is not a valid Python identifier." % self.label
+            )
```
Comment (chinmoy12c) [inline django/apps/config.py:40]: I guess this should be commented with the explanation.

```# app label must be a valid Python identifier```

Comment (charettes) [reply django/apps/config.py:40]: It feels like the code is pretty self documenting already.

Comment (felixxm) [reply django/apps/config.py:40]: I don't see any value in adding a comment that matches the error message.

File: django/db/migrations/state.py (hunk 2 above)
Comment (charettes) [reply django/db/migrations/state.py:238]: I agree if it doesn't cause to much pain in refactoring. It looks like `StateApps(Apps)` might be relying on other properties than `.label` and `.name`

Comment (felixxm) [reply django/db/migrations/state.py:238]: Yeah, I've checked and it's probably not worth to change this, thanks.

**Comment by felixxm** [comment]
@hramezani Thanks :+1: I pushed `AppConfigStub` cleanup to a separate commit.

**Comment by felixxm** [comment]
Merged in 110001d0bbbabe2a5b57b14a59bd0e4b71bf2712 and 8b2a30f6f16cb1f3538847954030d69da005bc7f.

**Comment by nemesisdesign** [comment]
Thanks @hramezani @felixxm :+1:

**Pull request closed without merging**